File: backend-python/src/nocobase_py/services/alert_store.py
"""R11 告警 SQLite 持久化 + 订阅.

为 AlertCollector 提供可选的持久化后端:
- 事件落盘 SQLite,重启可恢复
- 用户订阅(按 kind),只推送订阅的告警
- 订阅也持久化
- 不引入 SQLAlchemy 等重依赖,用 stdlib sqlite3

设计:
- 单表 `alerts` + 单表 `alert_subscriptions`
- 写盘为同步(简单可靠;数据量小写库快)
- 历史容量默认 10000,超过按时间淘汰最旧的 resolved
"""

from __future__ import annotations

import json
import logging
import sqlite3
import threading
import time
from collections.abc import Iterable
from pathlib import Path
from typing import Any


logger = logging.getLogger(__name__)


_SCHEMA_SQL = """
CREATE TABLE IF NOT EXISTS alerts (
    id TEXT PRIMARY KEY,
    kind TEXT NOT NULL,
    user_id TEXT,
    detail TEXT NOT NULL DEFAULT '{}',
    ts REAL NOT NULL,
    acked INTEGER NOT NULL DEFAULT 0,
    acked_by TEXT,
    acked_at REAL,
    resolved INTEGER NOT NULL DEFAULT 0,
    resolved_at REAL
);
CREATE INDEX IF NOT EXISTS idx_alerts_kind ON alerts(kind);
CREATE INDEX IF NOT EXISTS idx_alerts_ts ON alerts(ts);

CREATE TABLE IF NOT EXISTS alert_subscriptions (
    user_id TEXT NOT NULL,
    kind TEXT NOT NULL,
    created_at REAL NOT NULL,
    PRIMARY KEY (user_id, kind)
);
"""


class AlertStore:
    """告警事件 + 订阅的 SQLite 持久化存储.

    Usage:
        store = AlertStore(path="alerts.db")
        store.open()
        store.insert_alert({...})
        for ev in store.iter_alerts(limit=50):
            ...
        store.subscribe(user_id="u1", kind="rate_limit_exceeded")
    """

    def __init__(
        self,
        path: str | Path = "alerts.db",
        max_alerts: int = 10_000,
    ) -> None:
        self.path = Path(path)
        self.max_alerts = max_alerts
        self._lock = threading.Lock()
        self._conn: sqlite3.Connection | None = None

    def open(self) -> None:
        with self._lock:
            if self._conn is not None:
                return
            self.path.parent.mkdir(parents=True, exist_ok=True)
            self._conn = sqlite3.connect(str(self.path), check_same_thread=False)
            self._conn.row_factory = sqlite3.Row
            self._conn.executescript(_SCHEMA_SQL)
            self._conn.commit()
            logger.info("AlertStore opened: %s", self.path)

    def close(self) -> None:
        with self._lock:
            if self._conn is not None:
                self._conn.close()
                self._conn = None

    def _ensure_open(self) -> sqlite3.Connection:
        if self._conn is None:
            self.open()
        assert self._conn is not None
        return self._conn

    # ── 事件 CRUD ─────────────────────────────────────────────
    def insert_alert(self, ev: dict[str, Any]) -> None:
        conn = self._ensure_open()
        conn.execute(
            "INSERT OR REPLACE INTO alerts (id, kind, user_id, detail, ts, acked, acked_by, acked_at, resolved, resolved_at) "
            "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
            (
                ev["id"],
                ev["kind"],
                ev.get("user_id"),
                json.dumps(ev.get("detail") or {}),
                ev["timestamp"],
                int(bool(ev.get("acked"))),
                ev.get("acked_by"),
                ev.get("acked_at"),
                int(bool(ev.get("resolved"))),
                ev.get("resolved_at"),
            ),
        )
        conn.commit()
        self._trim_if_needed()

    def update_alert(
        self,
        event_id: str,
        *,
        acked: bool | None = None,
        acked_by: str | None = None,
        acked_at: float | None = None,
        resolved: bool | None = None,
        resolved_at: float | None = None,
    ) -> bool:
        conn = self._ensure_open()
        sets: list[str] = []
        vals: list[Any] = []
        if acked is not None:
            sets.append("acked = ?")
            vals.append(int(acked))
        if acked_by is not None:
            sets.append("acked_by = ?")
            vals.append(acked_by)
        if acked_at is not None:
            sets.append("acked_at = ?")
            vals.append(acked_at)
        if resolved is not None:
            sets.append("resolved = ?")
            vals.append(int(resolved))
        if resolved_at is not None:
            sets.append("resolved_at = ?")
            vals.append(resolved_at)
        if not sets:
            return False
        vals.append(event_id)
        cur = conn.execute(
            f"UPDATE alerts SET {', '.join(sets)} WHERE id = ?", vals
        )
        conn.commit()
        return cur.rowcount > 0

    def iter_alerts(
        self,
        limit: int = 100,
        *,
        include_resolved: bool = True,
        kinds: Iterable[str] | None = None,
    ) -> list[dict[str, Any]]:
        conn = self._ensure_open()
        sql = "SELECT * FROM alerts"
        where: list[str] = []
        params: list[Any] = []
        if not include_resolved:
            where.append("resolved = 0")
        if kinds:
            placeholders = ",".join("?" for _ in kinds)
            where.append(f"kind IN ({placeholders})")
            params.extend(list(kinds))
        if where:
            sql += " WHERE " + " AND ".join(where)
        sql += " ORDER BY ts DESC LIMIT ?"
        params.append(limit)
        rows = conn.execute(sql, params).fetchall()
        return [_row_to_dict(r) for r in rows]

    def count_alerts(self, *, include_resolved: bool = True) -> dict[str, int]:
        conn = self._ensure_open()
        sql = "SELECT kind, COUNT(*) AS n FROM alerts"
        if not include_resolved:
            sql += " WHERE resolved = 0"
        sql += " GROUP BY kind"
        return {r["kind"]: r["n"] for r in conn.execute(sql).fetchall()}

    def unresolved_count(self) -> int:
        conn = self._ensure_open()
        return conn.execute("SELECT COUNT(*) AS n FROM alerts WHERE resolved = 0").fetchone()["n"]

    def _trim_if_needed(self) -> None:
        """超过 max_alerts 时,删除最旧的 resolved 行."""
        conn = self._ensure_open()
        cur = conn.execute("SELECT COUNT(*) AS n FROM alerts")
        n = cur.fetchone()["n"]
        if n <= self.max_alerts:
            return
        excess = n - self.max_alerts
        conn.execute(
            "DELETE FROM alerts WHERE id IN ("
            "  SELECT id FROM alerts WHERE resolved = 1 ORDER BY ts ASC LIMIT ?"
            ")",
            (excess,),
        )
        conn.commit()

    # ── 订阅 CRUD ────────────────────────────────────────────
    def subscribe(self, user_id: str, kind: str) -> bool:
        """返回是否新增(True)或已存在(False)."""
        conn = self._ensure_open()
        cur = conn.execute(
            "INSERT OR IGNORE INTO alert_subscriptions (user_id, kind, created_at) VALUES (?, ?, ?)",
            (user_id, kind, time.time()),
        )
        conn.commit()
        return cur.rowcount > 0

    def unsubscribe(self, user_id: str, kind: str) -> bool:
        conn = self._ensure_open()
        cur = conn.execute(
            "DELETE FROM alert_subscriptions WHERE user_id = ? AND kind = ?",
            (user_id, kind),
        )
        conn.commit()
        return cur.rowcount > 0

    def subscriptions_for(self, user_id: str) -> list[str]:
        conn = self._ensure_open()
        rows = conn.execute(
            "SELECT kind FROM alert_subscriptions WHERE user_id = ? ORDER BY created_at",
            (user_id,),
        ).fetchall()
        return [r["kind"] for r in rows]

    def subscribers_for(self, kind: str) -> list[str]:
        """获取订阅了某 kind 的所有用户(用于推送过滤)."""
        conn = self._ensure_open()
        rows = conn.execute(
            "SELECT user_id FROM alert_subscriptions WHERE kind = ?",
            (kind,),
        ).fetchall()
        return [r["user_id"] for r in rows]


def _row_to_dict(r: sqlite3.Row) -> dict[str, Any]:
    return {
        "id": r["id"],
        "kind": r["kind"],
        "user_id": r["user_id"],
        "detail": json.loads(r["detail"] or "{}"),
        "timestamp": r["ts"],
        "acked": bool(r["acked"]),
        "acked_by": r["acked_by"],
        "acked_at": r["acked_at"],
        "resolved": bool(r["resolved"]),
        "resolved_at": r["resolved_at"],
    }


# ── 全局单例 ──────────────────────────────────────────────────────
_store: AlertStore | None = None


def get_store(
    path: str | Path | None = None,
    max_alerts: int | None = None,
    auto_attach: bool = True,
) -> AlertStore:
    """获取/初始化全局 AlertStore.

    第一次调用时可传 path/max_alerts;后续调用忽略.
    auto_attach=True 时自动附加到 AlertCollector(单例,只生效一次).
    """
    global _store
    if _store is None:
        if path is None:
            path = "alerts.db"
        if max_alerts is None:
            max_alerts = 10_000
        _store = AlertStore(path=path, max_alerts=max_alerts)
        _store.open()
        if auto_attach:
            try:
                from nocobase_py.services.alerts import get_collector

                get_collector().attach_store(_store)
            except Exception:  # noqa: BLE001
                logger.warning("自动 attach 到 AlertCollector 失败", exc_info=True)
        logger.info("全局 AlertStore 初始化完成")
    return _store


def reset_store_for_tests() -> None:
    """测试辅助:重置全局单例."""
    global _store
    if _store is not None:
        try:
            _store.close()
        except Exception:  # noqa: BLE001
            pass
    _store = None

File: backend-java/src/main/java/com/nocobase/alert/AlertStore.java
package com.nocobase.alert;

import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.annotation.PostConstruct;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Component;

import java.io.File;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * R11: 告警事件 + 订阅的 SQLite 持久化(JdbcTemplate).
 *
 * 通过 sqlite-jdbc 直连 SQLite 文件,不依赖 Spring Data JPA.
 * Spring 启动时自动建表.
 */
@Component
public class AlertStore {

    private static final String SCHEMA_SQL = """
            CREATE TABLE IF NOT EXISTS alerts (
                id TEXT PRIMARY KEY,
                kind TEXT NOT NULL,
                user_id TEXT,
                detail TEXT NOT NULL DEFAULT '{}',
                ts REAL NOT NULL,
                acked INTEGER NOT NULL DEFAULT 0,
                acked_by TEXT,
                acked_at REAL,
                resolved INTEGER NOT NULL DEFAULT 0,
                resolved_at REAL
            );
            CREATE INDEX IF NOT EXISTS idx_alerts_kind ON alerts(kind);
            CREATE INDEX IF NOT EXISTS idx_alerts_ts ON alerts(ts);

            CREATE TABLE IF NOT EXISTS alert_subscriptions (
                user_id TEXT NOT NULL,
                kind TEXT NOT NULL,
                created_at REAL NOT NULL,
                PRIMARY KEY (user_id, kind)
            );
            """;

    private final JdbcTemplate jdbc;
    private final ObjectMapper mapper = new ObjectMapper();
    private final String path;

    public AlertStore(
            JdbcTemplate jdbc,
            @Value("${nocobase.alerts.path:./alerts.db}") String path
    ) {
        this.jdbc = jdbc;
        this.path = path;
    }

    @PostConstruct
    public void init() throws Exception {
        // 确保父目录存在
        File f = new File(path);
        if (f.getParentFile() != null) {
            f.getParentFile().mkdirs();
        }
        // 给 JdbcTemplate 设置 SQLite 驱动(若未设置)
        try {
            Class.forName("org.sqlite.JDBC");
        } catch (ClassNotFoundException e) {
            throw new IllegalStateException("sqlite-jdbc 缺失", e);
        }
        // 逐条执行建表(SQLite JDBC 的 execute 对多条语句不友好)
        try (var conn = java.sql.DriverManager.getConnection("jdbc:sqlite:" + path);
             var stmt = conn.createStatement()) {
            for (String s : SCHEMA_SQL.split(";")) {
                String trimmed = s.trim();
                if (trimmed.isEmpty()) continue;
                stmt.execute(trimmed);
            }
        }
    }

    public String getPath() {
        return path;
    }

    // ── 事件 CRUD ─────────────────────────────────────────
    public void insertAlert(AlertEvent ev) {
        String detailJson;
        try {
            detailJson = mapper.writeValueAsString(ev.detail());
        } catch (Exception e) {
            detailJson = "{}";
        }
        try (var conn = java.sql.DriverManager.getConnection("jdbc:sqlite:" + path)) {
            var ps = conn.prepareStatement(
                    "INSERT OR REPLACE INTO alerts (id, kind, user_id, detail, ts, acked, acked_by, acked_at, resolved, resolved_at) "
                            + "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)"
            );
            ps.setString(1, ev.id());
            ps.setString(2, ev.kind());
            ps.setString(3, ev.userId());
            ps.setString(4, detailJson);
            ps.setDouble(5, ev.timestamp());
            ps.setInt(6, ev.acked() ? 1 : 0);
            ps.setString(7, ev.ackedBy());
            if (ev.ackedAt() != null) ps.setDouble(8, ev.ackedAt()); else ps.setNull(8, java.sql.Types.REAL);
            ps.setInt(9, ev.resolved() ? 1 : 0);
            if (ev.resolvedAt() != null) ps.setDouble(10, ev.resolvedAt()); else ps.setNull(10, java.sql.Types.REAL);
            ps.executeUpdate();
        } catch (Exception e) {
            throw new RuntimeException("insertAlert failed", e);
        }
    }

    public boolean updateAlert(
            String eventId,
            Boolean acked,
            String ackedBy,
            Double ackedAt,
            Boolean resolved,
            Double resolvedAt
    ) {
        StringBuilder sql = new StringBuilder("UPDATE alerts SET ");
        List<Object> params = new ArrayList<>();
        boolean first = true;
        if (acked != null) { sql.append(first ? "" : ", ").append("acked = ?"); params.add(acked ? 1 : 0); first = false; }
        if (ackedBy != null) { sql.append(first ? "" : ", ").append("acked_by = ?"); params.add(ackedBy); first = false; }
        if (ackedAt != null) { sql.append(first ? "" : ", ").append("acked_at = ?"); params.add(ackedAt); first = false; }
        if (resolved != null) { sql.append(first ? "" : ", ").append("resolved = ?"); params.add(resolved ? 1 : 0); first = false; }
        if (resolvedAt != null) { sql.append(first ? "" : ", ").append("resolved_at = ?"); params.add(resolvedAt); first = false; }
        if (first) return false;
        sql.append(" WHERE id = ?");
        params.add(eventId);
        try (var conn = java.sql.DriverManager.getConnection("jdbc:sqlite:" + path);
             var ps = conn.prepareStatement(sql.toString())) {
            for (int i = 0; i < params.size(); i++) ps.setObject(i + 1, params.get(i));
            return ps.executeUpdate() > 0;
        } catch (Exception e) {
            throw new RuntimeException("updateAlert failed", e);
        }
    }

    public List<Map<String, Object>> iterAlerts(int limit, boolean includeResolved, List<String> kinds) {
        StringBuilder sql = new StringBuilder("SELECT * FROM alerts");
        List<Object> params = new ArrayList<>();
        List<String> where = new ArrayList<>();
        if (!includeResolved) where.add("resolved = 0");
        if (kinds != null && !kinds.isEmpty()) {
            String placeholders = String.join(",", java.util.Collections.nCopies(kinds.size(), "?"));
            where.add("kind IN (" + placeholders + ")");
            params.addAll(kinds);
        }
        if (!where.isEmpty()) sql.append(" WHERE ").append(String.join(" AND ", where));
        sql.append(" ORDER BY ts DESC LIMIT ?");
        params.add(limit);
        try (var conn = java.sql.DriverManager.getConnection("jdbc:sqlite:" + path);
             var ps = conn.prepareStatement(sql.toString())) {
            for (int i = 0; i < params.size(); i++) ps.setObject(i + 1, params.get(i));
            var rs = ps.executeQuery();
            List<Map<String, Object>> result = new ArrayList<>();
            while (rs.next()) {
                Map<String, Object> row = new HashMap<>();
                row.put("id", rs.getString("id"));
                row.put("kind", rs.getString("kind"));
                row.put("user_id", rs.getString("user_id"));
                try {
                    row.put("detail", mapper.readValue(rs.getString("detail"), Map.class));
                } catch (Exception e) {
                    row.put("detail", new HashMap<>());
                }
                row.put("timestamp", rs.getDouble("ts"));
                row.put("acked", rs.getInt("acked") == 1);
                row.put("acked_by", rs.getString("acked_by"));
                row.put("acked_at", (Double) rs.getObject("acked_at"));
                row.put("resolved", rs.getInt("resolved") == 1);
                row.put("resolved_at", (Double) rs.getObject("resolved_at"));
                result.add(row);
            }
            return result;
        } catch (Exception e) {
            throw new RuntimeException("iterAlerts failed", e);
        }
    }

    public Map<String, Integer> countAlerts(boolean includeResolved) {
        String sql = "SELECT kind, COUNT(*) AS n FROM alerts"
                + (includeResolved ? "" : " WHERE resolved = 0")
                + " GROUP BY kind";
        try (var conn = java.sql.DriverManager.getConnection("jdbc:sqlite:" + path);
             var ps = conn.prepareStatement(sql);
             var rs = ps.executeQuery()) {
            Map<String, Integer> out = new HashMap<>();
            while (rs.next()) out.put(rs.getString("kind"), rs.getInt("n"));
            return out;
        } catch (Exception e) {
            throw new RuntimeException("countAlerts failed", e);
        }
    }

    public int unresolvedCount() {
        try (var conn = java.sql.DriverManager.getConnection("jdbc:sqlite:" + path);
             var ps = conn.prepareStatement("SELECT COUNT(*) AS n FROM alerts WHERE resolved = 0");
             var rs = ps.executeQuery()) {
            return rs.next() ? rs.getInt("n") : 0;
        } catch (Exception e) {
            throw new RuntimeException("unresolvedCount failed", e);
        }
    }

    // ── 订阅 CRUD ─────────────────────────────────────────
    public boolean subscribe(String userId, String kind) {
        try (var conn = java.sql.DriverManager.getConnection("jdbc:sqlite:" + path)) {
            var ps = conn.prepareStatement(
                    "INSERT OR IGNORE INTO alert_subscriptions (user_id, kind, created_at) VALUES (?, ?, ?)"
            );
            ps.setString(1, userId);
            ps.setString(2, kind);
            ps.setDouble(3, System.currentTimeMillis() / 1000.0);
            return ps.executeUpdate() > 0;
        } catch (Exception e) {
            throw new RuntimeException("subscribe failed", e);
        }
    }

    public boolean unsubscribe(String userId, String kind) {
        try (var conn = java.sql.DriverManager.getConnection("jdbc:sqlite:" + path);
             var ps = conn.prepareStatement(
                     "DELETE FROM alert_subscriptions WHERE user_id = ? AND kind = ?"
             )) {
            ps.setString(1, userId);
            ps.setString(2, kind);
            return ps.executeUpdate() > 0;
        } catch (Exception e) {
            throw new RuntimeException("unsubscribe failed", e);
        }
    }

    public List<String> subscriptionsFor(String userId) {
        try (var conn = java.sql.DriverManager.getConnection("jdbc:sqlite:" + path);
             var ps = conn.prepareStatement(
                     "SELECT kind FROM alert_subscriptions WHERE user_id = ? ORDER BY created_at"
             )) {
            ps.setString(1, userId);
            var rs = ps.executeQuery();
            List<String> out = new ArrayList<>();
            while (rs.next()) out.add(rs.getString("kind"));
            return out;
        } catch (Exception e) {
            throw new RuntimeException("subscriptionsFor failed", e);
        }
    }

    public List<String> subscribersFor(String kind) {
        try (var conn = java.sql.DriverManager.getConnection("jdbc:sqlite:" + path);
             var ps = conn.prepareStatement(
                     "SELECT user_id FROM alert_subscriptions WHERE kind = ?"
             )) {
            ps.setString(1, kind);
            var rs = ps.executeQuery();
            List<String> out = new ArrayList<>();
            while (rs.next()) out.add(rs.getString("user_id"));
            return out;
        } catch (Exception e) {
            throw new RuntimeException("subscribersFor failed", e);
        }
    }
}
